paddle/fluid/operators/controlflow/static_pylayer_op.cc

```diff
@@ -0,0 +1,279 @@
+              << ", outside_grad_name = " << outside_grad_name;
+      framework::Variable *outside_var =
+          parent_scope.FindVar(outside_grad_name);
+      if (outside_var == nullptr) {
+        continue;
+      }
+      framework::Variable *inside_var =
+          cur_scope.FindLocalVar(inside_grad_name);
+      if (inside_var == nullptr) {
+        assign_zero_outside_grads.emplace_back(outside_grad_name);
+        assign_zero_inputs.emplace_back(inputs[i]);
+        continue;
+      }
+      platform::DeviceContext *dev_ctx =
+          platform::DeviceContextPool::Instance().Get(place);
+      framework::VisitVarType(*inside_var,
+                              AssignFunctor(outside_var, *dev_ctx));
+    }
+  }
+};
+
+class StaticPyLayerBackwardInferShape : public framework::InferShapeBase {
+ public:
+  void operator()(framework::InferShapeContext *context) const override {
+    if (context->HasInputs(StaticPyLayerOp::kInputs) &&
+        context->HasOutputs(framework::GradVarName(StaticPyLayerOp::kInputs))) {
+      context->SetOutputsDim(framework::GradVarName(StaticPyLayerOp::kInputs),
+                             context->GetInputsDim(StaticPyLayerOp::kInputs));
+    }
+  }
+};
+
+class StaticPyLayerBackwardInferVarType : public framework::VarTypeInference {
+ public:
+  void operator()(framework::InferVarTypeContext *ctx) const override {
+    auto forward_input_size = ctx->InputSize(StaticPyLayerOp::kInputs);
+    auto backward_output_size =
+        ctx->OutputSize(framework::GradVarName(StaticPyLayerOp::kInputs));
+    PADDLE_ENFORCE_EQ(forward_input_size,
+                      backward_output_size,
+                      platform::errors::InvalidArgument(
+                          "input_size and output_size should be equal for "
+                          "static_pylayer_grad_op."));
+    for (size_t i = 0; i < backward_output_size; ++i) {
+      ctx->SyncTypeAndDataType(StaticPyLayerOp::kInputs,
+                               framework::GradVarName(StaticPyLayerOp::kInputs),
+                               i);
+    }
+  }
+};
+
+template <typename T>
+class StaticPyLayerBackwardMaker : public framework::SingleGradOpMaker<T> {
+ public:
+  using framework::SingleGradOpMaker<T>::SingleGradOpMaker;
+
+ protected:
+  void Apply(GradOpPtr<T> grad_op) const override {
+    grad_op->SetType("static_pylayer_grad");
+    grad_op->SetInput(StaticPyLayerOp::kInputs,
+                      this->Input(StaticPyLayerOp::kInputs));
+    grad_op->SetInput(framework::GradVarName(StaticPyLayerOp::kOutputs),
+                      this->OutputGrad(StaticPyLayerOp::kOutputs));
+    grad_op->SetInput(StaticPyLayerOp::kScope,
+                      this->Output(StaticPyLayerOp::kScope));
+
+    auto fwd_inputs = this->InputGrad(StaticPyLayerOp::kInputs, false);
+    grad_op->SetOutput(framework::GradVarName(StaticPyLayerOp::kInputs),
+                       fwd_inputs);
+    grad_op->SetBlockAttr("forward_block",
```

Comment:
```suggestion
    grad_op->SetBlockAttr("backward_block",
```
BackwardMaker里如果定义有 forward_block Attribute 感觉会比较奇怪